Failing to build due to "Pred not found in this scope" on newer versions of Emscripten

```
   Compiling skia-bindings v0.73.0 (/workspace/rust-skia/skia-bindings)
error[E0412]: cannot find type `_Pred` in this scope
  --> /workspace/rust-skia/target/wasm32-unknown-emscripten/release/build/skia-bindings-38874b9726951563/out/skia/bindings.rs:13:4396
   |
13 | ...uple__EnableUTypesCtor { pub _base : std___2__And < _Pred > , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tupl...
   |                                                        ^^^^^ not found in this scope
   |
help: you might be missing a type parameter
   |
13 | # [repr (C)] # [derive (Copy , Clone , Debug , Default , Eq , Hash , Ord , PartialEq , PartialOrd)] pub struct __BindgenBitfieldUnit < Storage > { storage : Storage , } impl < Storage > __BindgenBitfieldUnit < Storage > { # [inline] pub const fn new (storage : Storage) -> Self { Self { storage } } } impl < Storage > __BindgenBitfieldUnit < Storage > where Storage : AsRef < [u8] > + AsMut < [u8] >, { # [inline] pub fn get_bit (& self , index : usize) -> bool { debug_assert ! (index / 8 < self . storage . as_ref () . len ()) ; let byte_index = index / 8 ; let byte = self . storage . as_ref () [byte_index] ; let bit_index = if cfg ! (target_endian = "big") { 7 - (index % 8) } else { index % 8 } ; let mask = 1 << bit_index ; byte & mask == mask } # [inline] pub fn set_bit (& mut self , index : usize , val : bool) { debug_assert ! (index / 8 < self . storage . as_ref () . len ()) ; let byte_index = index / 8 ; let byte = & mut self . storage . as_mut () [byte_index] ; let bit_index = if cfg ! (target_endian = "big") { 7 - (index % 8) } else { index % 8 } ; let mask = 1 << bit_index ; if val { * byte |= mask ; } else { * byte &= ! mask ; } } # [inline] pub fn get (& self , bit_offset : usize , bit_width : u8) -> u64 { debug_assert ! (bit_width <= 64) ; debug_assert ! (bit_offset / 8 < self . storage . as_ref () . len ()) ; debug_assert ! ((bit_offset + (bit_width as usize)) / 8 <= self . storage . as_ref () . len ()) ; let mut val = 0 ; for i in 0 .. (bit_width as usize) { if self . get_bit (i + bit_offset) { let index = if cfg ! (target_endian = "big") { bit_width as usize - 1 - i } else { i } ; val |= 1 << index ; } } val } # [inline] pub fn set (& mut self , bit_offset : usize , bit_width : u8 , val : u64) { debug_assert ! (bit_width <= 64) ; debug_assert ! (bit_offset / 8 < self . storage . as_ref () . len ()) ; debug_assert ! ((bit_offset + (bit_width as usize)) / 8 <= self . storage . as_ref () . len ()) ; for i in 0 .. (bit_width as usize) { let mask = 1 << i ; let val_bit_is_set = val & mask == mask ; let index = if cfg ! (target_endian = "big") { bit_width as usize - 1 - i } else { i } ; self . set_bit (index + bit_offset , val_bit_is_set) ; } } } # [repr (C)] pub struct __BindgenUnionField < T > (:: core :: marker :: PhantomData < T >) ; impl < T > __BindgenUnionField < T > { # [inline] pub const fn new () -> Self { __BindgenUnionField (:: core :: marker :: PhantomData) } # [inline] pub unsafe fn as_ref (& self) -> & T { :: core :: mem :: transmute (self) } # [inline] pub unsafe fn as_mut (& mut self) -> & mut T { :: core :: mem :: transmute (self) } } impl < T > :: core :: default :: Default for __BindgenUnionField < T > { # [inline] fn default () -> Self { Self :: new () } } impl < T > :: core :: clone :: Clone for __BindgenUnionField < T > { # [inline] fn clone (& self) -> Self { * self } } impl < T > :: core :: marker :: Copy for __BindgenUnionField < T > { } impl < T > :: core :: fmt :: Debug for __BindgenUnionField < T > { fn fmt (& self , fmt : & mut :: core :: fmt :: Formatter < '_ >) -> :: core :: fmt :: Result { fmt . write_str ("__BindgenUnionField") } } impl < T > :: core :: hash :: Hash for __BindgenUnionField < T > { fn hash < H : :: core :: hash :: Hasher > (& self , _state : & mut H) { } } impl < T > :: core :: cmp :: PartialEq for __BindgenUnionField < T > { fn eq (& self , _other : & __BindgenUnionField < T >) -> bool { true } } impl < T > :: core :: cmp :: Eq for __BindgenUnionField < T > { } pub type std___2___enable_if_t = u8 ; pub type std___2_integral_constant_value_type < _Tp > = _Tp ; pub type std___2_integral_constant_type = u8 ; pub type std___2_true_type = u8 ; pub type std___2_false_type = u8 ; pub type std___2__If = u8 ; pub type std___2_conditional_type < _If > = _If ; pub type std___2___conditional_t = u8 ; pub type std___2__And < _Pred > = _Pred ; pub type std___2__Or = u8 ; # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2___tuple_impl { pub _address : u8 , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple { pub __base_ : std___2_tuple__BaseT , } pub type std___2_tuple__BaseT = std___2___tuple_impl ; # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__IsThisTuple { pub _base : std___2_false_type , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__EnableUTypesCtor<_Pred> { pub _base : std___2__And < _Pred > , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__EnableCtorFromUTypesTuple { pub _base : std___2_false_type , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__CtorPredicateFromPair { pub _base : std___2_false_type , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__EnableCtorFromPair { pub _base : std___2_tuple__CtorPredicateFromPair , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__NothrowConstructibleFromPair { pub _base : std___2_tuple__CtorPredicateFromPair , } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_tuple__BothImplicitlyConvertible { pub _base : std___2_false_type , } pub type std___2_string = std___2_basic_string ; pub type std___2_streamoff = :: core :: ffi :: c_longlong ; pub type std___2_streampos = std___2_fpos < __mbstate_t > ; # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_input_iterator_tag { pub _address : u8 , } # [test] fn bindgen_test_layout_std___2_input_iterator_tag () { assert_eq ! (:: core :: mem :: size_of :: < std___2_input_iterator_tag > () , 1usize , concat ! ("Size of: " , stringify ! (std___2_input_iterator_tag))) ; assert_eq ! (:: core :: mem :: align_of :: < std___2_input_iterator_tag > () , 1usize , concat ! ("Alignment of " , stringify ! (std___2_input_iterator_tag))) ; } # [repr (C)] # [derive (Debug , Copy , Clone)] pub struct std___2_forward_iterator_tag { pub _address : u8 , } # [test] fn bindgen_test_layout_std___2_forward_iterator_tag () { assert_eq ! (:: core :: mem :: size_of :: < std___2_forward_iterator_tag > () , 1usize , concat ! ("Size of: " , stringify ! (std___2_forward_iterator_tag))) ; assert_eq ! (:: core :: mem :: align_of :: < std___2_forward_iterator_tag > () , 1usize , concat ! ("Alignment of " , stringify ! (std___2_forward_iterator_tag))) ; }
<truncated>
```

Environment:
- Standard Rust docker container (Debian Linux Bookworm)
- rust-skia 0.73.0 (can also repro on 0.72.0)
- Rust 1.78.0 (can also repro on 1.77.2)
- LLVM 19 (can also repro on LLVM 18)
- target is wasm32-unknown-emscripten

Looks like a similar issue as:
https://github.com/rust-skia/rust-skia/issues/571

Is there something else that needs to be allowlisted?
https://github.com/rust-skia/rust-skia/pull/573

cc @pragmatrix

--- a/README.md
+++ b/README.md
@@ -217,7 +217,7 @@ Compilation to iOS is supported on macOS targeting the iOS simulator (`--target 
 
 ### For WebAssembly
 
-Install `emscripten` version 3.1.57 or superior. In the examples below, we assume
+Install `emscripten` version 3.1.57 or superior and make sure that llvm / clang 16+ is installed. In the examples below, we assume
 `emsdk` version `3.1.57` was installed with [asdf](http://asdf-vm.com/).
 
 Build with the `wasm32-unknown-emscripten` target (`wasm32-unknown-unknown` is

--- a/README.md
+++ b/README.md
@@ -217,14 +217,14 @@ Compilation to iOS is supported on macOS targeting the iOS simulator (`--target 
 
 ### For WebAssembly
 
-Install `emscripten` version 3.1.1 or superior. In the examples below, we assume
-`emsdk` version `3.1.1` was installed with [asdf](http://asdf-vm.com/).
+Install `emscripten` version 3.1.57 or superior. In the examples below, we assume
+`emsdk` version `3.1.57` was installed with [asdf](http://asdf-vm.com/).
 
 Build with the `wasm32-unknown-emscripten` target (`wasm32-unknown-unknown` is
 unsupported because it is [fundamentally incompatible with linking C code](https://github.com/rustwasm/team/issues/291#issuecomment-645482430):
 
 ```bash
-export EMSDK=~/.asdf/installs/emsdk/3.1.1
+export EMSDK=~/.asdf/installs/emsdk/3.1.57
 export EMCC_CFLAGS="-s ERROR_ON_UNDEFINED_SYMBOLS=0"
 
 cargo build --target wasm32-unknown-emscripten
@@ -239,7 +239,7 @@ In `EMCC_CFLAGS`, `-s ERROR_ON_UNDEFINED_SYMBOLS` is a
 If you want to enable WebGL, you will also have to set `MAX_WEBGL_VERSION=2`:
 
 ```bash
-export EMSDK=~/.asdf/installs/emsdk/3.1.1
+export EMSDK=~/.asdf/installs/emsdk/3.1.57
 export EMCC_CFLAGS="-s ERROR_ON_UNDEFINED_SYMBOLS=0 -s MAX_WEBGL_VERSION=2"
 
 cargo build --target wasm32-unknown-emscripten --features gl
@@ -256,7 +256,7 @@ architecture, and can be retrieved with `brew info binutils`. Here is an
 example for Apple silicon:
 
 ```bash
-export EMSDK=~/.asdf/installs/emsdk/3.1.1
+export EMSDK=~/.asdf/installs/emsdk/3.1.57
 export EMCC_CFLAGS="-s ERROR_ON_UNDEFINED_SYMBOLS=0"
 export PATH="/opt/homebrew/opt/binutils/bin:$PATH"
 

--- a/skia-bindings/build_support/skia_bindgen.rs
+++ b/skia-bindings/build_support/skia_bindgen.rs
@@ -444,6 +444,8 @@ const OPAQUE_TYPES: &[&str] = &[
     "skia_private::THashMap",
     // m121:
     "skgpu::MutableTextureState",
+    // emscripten: Uses SkLRUCache (which is blocklisted)
+    "skia::textlayout::ParagraphCache",
 ];
 
 const BLOCKLISTED_TYPES: &[&str] = &[

--- a/skia-bindings/build_support/skia_bindgen.rs
+++ b/skia-bindings/build_support/skia_bindgen.rs
@@ -423,6 +423,8 @@ const OPAQUE_TYPES: &[&str] = &[
     "std::tuple",
     // Homebrew macOS LLVM 13
     "std::tuple_.*",
+    // Since 3.1.57 of the emsdk: <https://github.com/rust-skia/rust-skia/issues/975>
+    "std::__2::tuple.*",
     // m93: private, exposed by Paint::asBlendMode(), fails layout tests.
     "skstd::optional",
     // m100

--- a/skia-bindings/src/paragraph.cpp
+++ b/skia-bindings/src/paragraph.cpp
@@ -129,6 +129,10 @@ extern "C" {
         self->~ParagraphCache();
     }
 
+    void C_ParagraphCache_turnOn(ParagraphCache* self, bool value) {
+        self->turnOn(value);
+    }
+
     int C_ParagraphCache_count(ParagraphCache* self) {
         return self->count();
     }

--- a/skia-safe/src/interop/string.rs
+++ b/skia-safe/src/interop/string.rs
@@ -73,7 +73,19 @@ impl AsStr for SkString {
     }
 }
 
-impl AsStr for sb::std_string_view {
+// Since 3.1.57 of the emsdk: <https://github.com/rust-skia/rust-skia/issues/975>
+#[cfg(not(target_arch = "wasm32"))]
+mod std_cpp {
+    pub type StringView = skia_bindings::std_string_view;
+    pub type String = skia_bindings::std_string;
+}
+#[cfg(target_arch = "wasm32")]
+mod std_cpp {
+    pub type StringView = skia_bindings::std___2_string_view;
+    pub type String = skia_bindings::std___2_string;
+}
+
+impl AsStr for std_cpp::StringView {
     fn as_str(&self) -> &str {
         let slice = unsafe {
             let mut size = 0;
@@ -84,7 +96,7 @@ impl AsStr for sb::std_string_view {
     }
 }
 
-impl AsStr for sb::std_string {
+impl AsStr for std_cpp::String {
     fn as_str(&self) -> &str {
         let slice = unsafe {
             let mut size = 0;

--- a/skia-safe/src/modules/paragraph/paragraph_cache.rs
+++ b/skia-safe/src/modules/paragraph/paragraph_cache.rs
@@ -34,7 +34,7 @@ impl ParagraphCache {
     }
 
     pub fn turn_on(&mut self, value: bool) {
-        self.native_mut().fCacheIsOn = value
+        unsafe { sb::C_ParagraphCache_turnOn(self.native_mut(), value) }
     }
 
     pub fn count(&mut self) -> i32 {

--- a/wasm-example/src/main.rs
+++ b/wasm-example/src/main.rs
@@ -1,8 +1,9 @@
+use std::boxed::Box;
+
 use skia_safe::{
     gpu::{self, gl::FramebufferInfo, DirectContext},
     Color, Paint, PaintStyle, Surface,
 };
-use std::boxed::Box;
 
 extern "C" {
     pub fn emscripten_GetProcAddress(
@@ -51,7 +52,7 @@ fn init_gl() {
 /// This needs to be done once per WebGL context.
 fn create_gpu_state() -> GpuState {
     let interface = skia_safe::gpu::gl::Interface::new_native().unwrap();
-    let context = skia_safe::gpu::DirectContext::new_gl(interface, None).unwrap();
+    let context = skia_safe::gpu::direct_contexts::make_gl(interface, None).unwrap();
     let framebuffer_info = {
         let mut fboid: gl::types::GLint = 0;
         unsafe { gl::GetIntegerv(gl::FRAMEBUFFER_BINDING, &mut fboid) };

